### PR TITLE
added various std traits for `PyBackedStr` and `PyBackedBytes`

### PR DESCRIPTION
**Pull request by Icxolu**
Adds `Clone`, `Debug`, `PartialEq`, `Eq`, `PartialOrd`, `Ord` and `Hash` implementation for `PyBackedBytes` and `PyBackedStr`, and `Display` for `PyBackedStr`. The implementations just delegate to `&[u8]` and `&str` respectively. `PartialEq` and `PartialOrd` are also implemented between `Self` and `Deref::Target` respectively.

### CODE REVIEW

**Review by LilyFoote — APPROVED** [review]
Looks good to me. Do we need tests?

**Comment by Icxolu** [comment]
> Do we need tests?

I was wondering about that as well. I'll try to come up with a few later.

**Comment by davidhewitt** [comment]
I wonder, can we ship these in 0.21.1?

Theoretically adding trait implementations can be breaking, but these types are so new and the traits so basic I can't really see them being a problem.

I would definitely like to ship the Send / Sync implementations in 0.21.1.

File: src/pybacked.rs [hunk 1]
```diff
@@ -79,13 +88,15 @@ impl FromPyObject<'_> for PyBackedStr {
 /// A wrapper around `[u8]` where the storage is either owned by a Python `bytes` object, or a Rust `Box<[u8]>`.
 ///
 /// This type gives access to the underlying data via a `Deref` implementation.
+#[derive(Clone)]
```
Comment (adamreichold) [inline src/pybacked.rs:91]: Isn't the derived `Clone` impl incorrect for the `Rust` variant? Cloning `Box<[u8]>` will yield a new allocation for `data` will still point to the old one meaning this is a use-after-free in waiting?

So we either need to manually fix up the `data` pointer or use a shared ownership pointer for the Rust side as well, e.g. `Arc<[u8]>`.

Comment (adamreichold) [reply src/pybacked.rs:91]: (And so yes, these do need tests. ;-))

Comment (davidhewitt) [reply src/pybacked.rs:91]: Yes I was thinking similarly about the `Box`, an `Arc` seems reasonable to me.

I also half wonder if it was a mistake to have the Rust variant here given the name `PyBackedBytes`. I added it to support bytearrays in the same way we already do for `Cow<[u8]>` but on further reflection it does seem a bit add odds with the name that we do the copy.

I think probably it's ok as I would expect most uses to be with bytes, but we might want to document this copy-from-bytearray more clearly and justify for sake of safety.

Comment (Icxolu) [reply src/pybacked.rs:91]: > Isn't the derived `Clone` impl incorrect for the `Rust` variant? Cloning `Box<[u8]>` will yield a new allocation for `data` will still point to the old one meaning this is a use-after-free in waiting?

Big oops, thanks for catching.

Using `Arc<[u8]>` sounds good to me aswell, that way this stays a cheap clone either way.

**Comment by Icxolu** [comment]
I've added the change to `Arc<[u8]>` and tests for the traits impls. I also added a few more `PartialEq` impls that came useful while writing the tests.

> I wonder, can we ship these in 0.21.1? Theoretically adding trait implementations can be breaking, but these types are so new and the traits so basic I can't really see them being a problem.

I agree, since these types are so new, I think the chances of anyone relying on these impls _not_ to be there are very low. Adding them in 0.21.1 should be fine I would say.

**Review by davidhewitt — APPROVED** [review]
Thanks, looks good to me! I'll start preparing 0.21.1 a bit later today 👍

**Pull request merged**